### ISSUE TITLE
Release 2.6.10 nested srai fails to preserve star values

### ISSUE THREAD

**Issue by tringate** [Program-O/Program-O#388]
The following test cases demonstrate the failure of a nested srai to preserve the values of the STAR.

These are new test cases for the Program-O project and my intent was to find a simple test case
that Dave can use while working on this bug.  The logic of the test cases appears to be valid to me to test this failure.

Issue is the loss of the STAR value when nesting SRAIs which use the star values

```

<!-- test case77 -->
<!-- test if nested SRAI causes any change * value -->
<!-- do 2 srai's using a star value, on return check if both star values survived the nested srai by setting the STAR value into a predicate before returning -->
<category><pattern>TESTCASE77</pattern><template>
<think>
<srai>SRAISTAR fu</srai>
</think>
<condition name="testname">
	<li value="fu">
		<condition name="testname1">
		<li value="bar"><set name="testname1">*** PASSED ***</set></li>
		<li>*** FAILED ***</li>
		</condition></li>
	<li>*** FAILED *** <get name="testname"/> <get name="testname1"/></li>
</condition>
</template></category>

<category><pattern>SRAISTAR *</pattern><template>
<srai>SRAISTAR1 bar</srai>
<condition name="*">
	<li value="fu"><set name="testname"><star/></set></li>
	<li><set name="testname">*** NOT FU ***</set></li>
</condition>
</template></category>

<category><pattern>SRAISTAR1 *</pattern><template>
<condition name="*">
	<li value="bar"><set name="testname1"><star/></set></li>
	<li><set name="testname1">*** NOT BAR ***</set></li>
</condition>
</template></category>

<!-- test case78 -->
<!-- test if nested SRAI causes any change * value -->
<!-- do 3 srai's using a star value, on return check if both star values survived the nested srai by setting the STAR value into a predicate before returning -->
<category><pattern>TESTCASE78</pattern><template>
<think>
<srai>SRAISTAR2 fubar</srai>
</think>
<condition name="testname">
	<li value="fubar">
		<condition name="testname1">
		<li value="bar"><set name="testname1">*** PASSED ***</set></li>
		<li>*** FAILED ***</li>
		</condition></li>
	<li>*** FAILED *** <get name="testname"/> <get name="testname1"/> <get name="testname2"/></li>
</condition>
</template></category>

<category><pattern>SRAISTAR2 *</pattern><template>
<srai>SRAISTAR bar</srai>
<condition name="*">
	<li value="fubar"><set name="testname2"><star/></set></li>
	<li><set name="testname2">*** NOT FUBAR ***</set></li>
</condition>
</template></category>

`

**Comment by OverlanAI** [comment]
good catch! I was suspecting the nested stuff was not WAI but in my 50k worth of cats and a lowly 7k of working SRAI calls, and after looking in what i have i found just a small fraction that stack it like that.
So thats a good catch and hard to know that it may be wrong.
Tnx (I think your right ).

**Comment by xhtmlguru** [comment]
Confirmed issue.

**Comment by OverlanAI** [comment]
Seems like a solid patch here.
Added both of your fixes manually using my Huge RUE bot set and am seeing the effect.

nice job and tnx.

**Comment by tringate** [comment]
I have confirmed this issue is resolved. I also have rewritten the test case as it was flawed in how it was testing this function. It found the bug by accident I believe.

**Comment by tringate** [comment]
Patrick, I moved this back to in progress based on your comments and the test results of the current test case AIML testing

**Comment by OverlanAI** [comment]
I noticed that there is a marked up-tick in cpu usage after these patches. Before patching and with a nearly streamed chat on an 8 core workstation we were seeing around 5 to 10%, After the last few updates it can top out around 40%.
I am just mentioning this so you can keep an eye on how hard this can hit mysql and apache over all.
Keep in mind that i use a huge aiml set and chats are almost always at or near a streaming to just one bot.
Anyway might want to keep an eye on how hard the changes effect performance.
tnx (just a mention not a big issue yet).

**Comment by OverlanAI** [comment]
Umm, yes but now that i found a workaround so i could fill SRAI table, i find it digs in too deep to the point the bot stops talking. 
I know (and before you get that look) there are 1k things that could cause it.
My point is i can revert to release and the same set always has something to say.
might just be things were never right in the first place. 
tnx

**Comment by OverlanAI** [comment]
Good that helps, At least it is going it a lot deeper into the reductions that it was before the patch.
What i mostly get here is it will do a few rounds of chat then it bot will say nothing back no matter what the topic is but its real random and does not look like its stuck in a loop. After a few seconds i will say something else to it and it will do a few more rounds of chat.
Anyway im not so sure that it is the AMIL. Im still digging around to see if i can cause the problem each time vrs random using a complex set.
I can use the same topic saying the same things to it and sometimes its WAI and other times it is not.
Mean time i am doing my best to check all 6 levels of reductions it has but man thats a pain.
Tnx for the info, that help to know what it should be doing.

**Comment by OverlanAI** [comment]
Im not so sure about this patch right now. After getting the SRAI problem working correctly (that problem pre-dated this) may be why it was looking like one thing but was not.
Now i seem to be getting all sorts of * value issues that look like what this patch was meant to fix.
Anyway, it does process the nested SRAI's but something i wonky with the * values that it passes.
At least now the SRAI IDs match to the AIML IDs correctly now but that patch is not in dev just yet as that section is getting a face lift right now.
Before the SRAI id's did not match the AMIL IDs so would have made one problem look like something else than a patch was added to the something else problem without fixing the root problem.
Anyway after the SRAI face-lift we should re visit the patches that were added after the fact.
TNX

**Comment by OverlanAI** [comment]
all yours, we tried to help

**Comment by Program-O** [comment]
I have ran the AIML on the latest version and i am seeing a failure...
```Seeker: TESTCASE78
testbot: *** FAILED *** *** NOT FU *** bar fubar ```
I will dig into this further over the weekend

**Comment by tringate** [comment]
I just installed release 2.6.11 and ran the test cases against the install here is the results
Seeker: QUICKTEST
root: 
*** TEST CASES ARE DOCUMENTED IN THE AIML FILE *** 
test case 001 - OK 
test case 002 - O K 
test case 003 - OK 
test case 004 - OK 
test case 005 - OK 
test case 006 - OK 
test case 007 - OK 
test case 008 - OK 
test case 009 - OK 
test case 010 - OK 
test case 011 - OK 
test case 012 - OK 
test case 013 - OK 
test case 014 - OK 
test case 015 - OK 
test case 016 - OK 
test case 017 - OK 
test case 018 - OK 
test case 019 - OK 
test case 020 - OK 
test case 021 - OK 
test case 022 - OK 
test case 023 - OK 
test case 024 - OK 
test case 025 - OK 
test case 026 - OK 
test case 027 - OK 
test case 028 - OK 
test case 029 - OK 
test case 030 - OK 
test case 031 - OK 
test case 032 - OK 
test case 033 - OK 
test case 034 - OK 
test case 035 - OK 
test case 036 - OK 
test case 037 - OK 
test case 038 - OK 
test case 039 - OK 
test case 040 - OK 
test case 041 - OK 
test case 042 - OK 
test case 043 - OK 
test case 044 - OK 
test case 045 - OK 
test case 046 - OK 
test case 047 - OK 
test case 048 - OK 
test case 049 - OK 
test case 050 - OK 
test case 051 - OK 
test case 052 - OK 
test case 053 - OK 
test case 054 - OK 
test case 055 - OK 
test case 056 - OK 
test case 057 - OK 
test case 058 - OK 
test case 059 - OK 
test case 060 - OK 
test case 061 - OK 
test case 062 - OK 
test case 063 - OK 
test case 064 - OK 
test case 065 - OK 
test case 066 - OK 
test case 067 - OK 
test case 068 - OK 
test case 069 - OK 
test case 070 - OK 
test case 071 - OK 
test case 072 - OK 
test case 073 - OK 
test case 074 - OK 
test case 075 - OK 
test case 076 - OK 
test case 077 - OK 
test case 078 - OK 
test case 079 - OK 
test case 080 - OK 
test case 081 - OK 
test case 082 - OK
 
This is a clean run, so I would have to say all is OK.

Once I test against the "genie" version of program-o I will know for sure if this is still an issue.
I will add one comment about this issue.  I have made it a practice to save any variable passed in a srai in a real variable if I am going to do  nested SRAI's.  I always use the variable names of test,test1,test2, tec. so avoid making tons of oddball variables to clutter up the database.

**Comment by tringate** [comment]
I have updated my production "genie" code to release 2.7.0 2018-11-04 and have several nested "SRAI" categories that call them self as part of a "list" feature I have implemented.

I looked into my AIML and see I have converted all "srai" usage to always save any passed variable if they were going to be used later in the chain of "srai" execution.  Because of that step, my code does not depend on any star variable being preserved.  In my test cases this has not been done and they work properly and preserve the star valuables.
